### ISSUE TITLE
TV Shows added to IMDB Top Movies list?

### ISSUE THREAD

**Issue by BEisem** [Taxel/PlexTraktSync#228]
When I do a full sync, I'm seeing several entries in the log like the following:

```
2021-04-11 19:21:35,765 INFO[PlexTraktSync]:Show [St. Elsewhere (1982)]: s02e09 added to list IMDB: Top Rated Movies
2021-04-11 19:21:35,786 INFO[PlexTraktSync]:Show [St. Elsewhere (1982)]: s02e15 added to list IMDB: Top Rated Movies
```

The problem is, it is indicating that a TV show episode is being added to a list called IMDB Top Rated Movies.

**Comment by glensc** [comment]
@BEisem Please include version info in your bugreport!

But other than that, I don't understand what are you reporting here:
1. what is the problem with what it's doing now?
2. what is the expected outcome?

I'm not familiar with the feature myself.

**Comment by BEisem** [comment]
Sorry, I don't know how to determine the version. I pulled this from the "main" branch. Is that sufficient? How else can I determine the version I'm using?

As far as the problem... the list is called IMDB Top Rated Movies, but it is adding TV show episodes to the list. I guess I would expect it to add movies to a movie list, not TV episodes.

I'm not familiar with the feature either, I just noticed the entries in the logs, so I figured if it was going to add things to a movie list, it should be adding movies, not TV episodes.

**Comment by glensc** [comment]
> Sorry, I don't know how to determine the version. I pulled this from the "main" branch. Is that sufficient? How else can I determine the version I'm using?

Third time already pointing you to this info:
- https://github.com/Taxel/PlexTraktSync/pull/195

I do not know what means "pulled from the main branch" in your case. if it's `git pull`, then `git describe --tags` should print the closest tag. if you used zip download, you should have used zip for tag rather than a branch.

**Comment by glensc** [comment]
regarding the issue at hand: ok, now I see what you mean.

if you don't use the feature you can turn it off, it's either "liked_lists" or "watchlist":
- https://github.com/Taxel/PlexTraktSync/blob/22a1ef60cab82a7bf02d6b04b7d2cb0551b894f4/config.default.json#L4-L5

**Comment by BEisem** [comment]
Ok, previously I was downloading the zip file from the "main" branch, but now I understand how to use `git clone` with a tag. I have cloned 0.6.44 (currently the latest) and tested this, and it is still occurring.

I know I could turn off the feature, but as a test, I believe it should be fixed anyway, in case someone does want to use the feature.

**Comment by glensc** [comment]
I don't argue it shouldn't be fixed :)

**Comment by twolaw** [comment]
The script is mixing trakt ids from Movies and TV episodes. That's the reason.

Some trakt ids are identicals such as **1034** corresponding to both :

- St. Elsewhere (1982) s02e09
- Raging Bull (a movie in IMDB top 250)

id **1040** corresponding to both :

- St. Elsewhere (1982) s02e15
- It's a Wonderfull Life (a movie in IMDB top 250)

I didn't dig further into the code yet.
When searching an item from its trakt id, the script should specify the type `shows` or `movies`.

**Comment by glensc** [comment]
@twolaw can you point to any code you looked at?

at least our search_by_id trakt method matches correct media type:
- https://github.com/Taxel/PlexTraktSync/blob/06ee4af54d47072903402de112c3caf59db69759/plex_trakt_sync/trakt_api.py#L213-L221

however, our extra check is no longer necessary there, as media_type is passed as an argument. haven't yet cleaned up.

**Comment by twolaw** [comment]
I don't think it's in this search_by_id function.
I think it's when script goes through every episode, it check if its id is in any lists to do the match with Plex id and add to Plex list. At this moment the **type** check may be missing.
I say this from my mind, I will look the code later.

**Comment by glensc** [comment]
@twolaw ah, I thought you already looked at the code.

**Comment by glensc** [comment]
pm/tm/pe/te are gone now (mostly):
- https://github.com/Taxel/PlexTraktSync/pull/254

**Comment by twolaw** [comment]
@glensc why only watchlist_movies and no watchlist_shows ?
https://github.com/Taxel/PlexTraktSync/blob/c678c2c0c32084a1175d35f88d59ba2a0aefca70/plex_trakt_sync/commands/sync.py#L166-L172

(just asking, maybe new issue is needed)

**Comment by glensc** [comment]
@twolaw I have not refactored the watchlist thing. so it's just carried over preserving original behavior/logic. I have not touched that part at all, so I can't comment as I don't understand it.

ps: the latest code is in `0.8.x` branch.

ps2: here's an overview of the branches: https://github.com/Taxel/PlexTraktSync/issues/162#issuecomment-821784403

**Comment by twolaw** [comment]
Yes I get it now.

As far as I understand it :

- A **watchlist** will only contain movies because the script filters to get only movies (trakt can put anything in [a watchlist](https://trakt.docs.apiary.io/#reference/users/watchlist/get-watchlist)). That's quite fair because nobody would create a watchlist of tv episodes, that's useless.

- A **custom list** can contain movies and episodes. The script filters the trakt response to remove shows and seasons because Plex cannot put a show or a season in a list.

The script stores watchlist and custom list as `TraktList` object and removes the attribute movie or episode when building the list. That's the problem :
https://github.com/Taxel/PlexTraktSync/blob/c678c2c0c32084a1175d35f88d59ba2a0aefca70/plex_trakt_sync/trakt_list_util.py#L17

Usefull links:
#124
#134

**Comment by twolaw** [comment]
@glensc why is there `listutil.addPlexItemToLists(me.trakt_id, me.plex.item)` for tv and not for movies ?
In 0.6.x the logic has changed and it looks like addPlexItemToLists for movies have been removed https://github.com/Taxel/PlexTraktSync/commit/5d06e61dc99ac26419c6d36a1fc3b362bd3133a3
Can you help me understand ?

**Comment by glensc** [comment]
@twolaw I guess I was confused with the listutil line below when refactoring:
- https://github.com/Taxel/PlexTraktSync/blob/cdb1aae5369485b8c73ac7a22efb486ed2b857a1/plex_trakt_sync/commands/sync.py#L141-L142

Submitted a possible fix: https://github.com/Taxel/PlexTraktSync/pull/287

**Comment by glensc** [comment]
@twolaw https://github.com/Taxel/PlexTraktSync/pull/287 was merged, does this make https://github.com/Taxel/PlexTraktSync/issues/228 solved?

**Comment by twolaw** [comment]
No, I'm waiting an answer to https://github.com/trakt/api-help/discussions/287 then I'll write a fix.

**Comment by glensc** [comment]
@twolaw I don't understand that question there. you should include URLs or API queries where you got those results. to be clearer. and also, what is unicity? 

as I answered here before, each domain (movie,show,episode,actor,director, etc) has its own id, so across the domains, it's expected to have id overlap. you should match with media_type in the queries. as for example for search I already provided an example:
- https://github.com/Taxel/PlexTraktSync/issues/228#issuecomment-822497082

**Comment by twolaw** [comment]
As I said it is a clearly defined bug in trakt_list_util (line 17 and more) it is not a problem about trakt queries or trakt answers.

**Comment by glensc** [comment]
Fine, I'll leave you to it. I don't think I will look at this anytime soon myself.

ps: you should share links to resources you've researched, rather than some file somewhere some line. helps others to join the brainstorming and maybe even future you.

**Comment by twolaw** [comment]
@BEisem can you try again **IMDB: Top Rated Movies** list sync with this fix #293 and tell me if it's of for you ?

**Comment by BEisem** [comment]
@twolaw can you give me instructions on how to sync a specific list? I've never used this feature so I don't know how it works, and I can't find any documentation. Otherwise I'd be happy to test it...

**Comment by BEisem** [comment]
Ok, that's what I thought ... but I'm getting errors. Do I have to have a playlist in Plex named IMDB: Top Rated Movies?

```
INFO: Syncing with Plex BlaEis and Trakt binkatl
INFO: Filtering Library: Stand-Up Comedy
INFO: Syncing TV=True, Movies=True
INFO: Loaded Trakt lists in 14.1 seconds
INFO: Downloaded List Rotten Tomatoes: Best of 2019
INFO: Downloaded List IMDB: Top Rated Movies
INFO: Server version 1.22.3.4392-d7c624def updated at: 2021-04-21 10:20:45
INFO: Recently added: [<Movie:386104:F.E.A.R.>, <Season:386081:Phil-of-the-Future-s2>, <Season:386058:Even-Stevens-s3>, <Season:386035:Even-Stevens-s2>, <Season:386014:Even-Stevens-s1>]
Processing Stand-Up Comedy  [##############----------------------]  130/323  00:00:02WARNING: Skipping <imdb:tt2243073:<Movie:365451:Hannah-Gadsby:-Kiss->>: Not found on Trakt
Processing Stand-Up Comedy  [#########################-----------]  229/323  00:00:01WARNING: Skipping <imdb:tt6140294:<Movie:365744:The-Nasty-Show-Hoste>>: Not found on Trakt
Processing Stand-Up Comedy  [####################################]  323/323          
INFO: Stand-Up Comedy processed in 4.3 seconds
ERROR: Playlist Rotten Tomatoes: Best of 2019 not found, so it could not be deleted. Actual playlists: [<Playlist:/playlists/282615:70's-Comedies>, <Playlist:/playlists/280507:80s-Comedy>, <Playlist:/playlists/282612:90's-Comedies>, <Playlist:/playlists/282613:2000's-Comedies>, <Playlist:/playlists/261465:All-Music>, <Playlist:/playlists/261476:All-Music>, <Playlist:/playlists/261466:Recently-Added>, <Playlist:/playlists/261475:Recently-Added>, <Playlist:/playlists/261467:Recently-Played>, <Playlist:/playlists/261477:Recently-Played>, <Playlist:/playlists/128363:Action>, <Playlist:/playlists/141517:Blake-and-Diane>, <Playlist:/playlists/372604:Blake-and-Diane-Movi>, <Playlist:/playlists/128362:Comedy>, <Playlist:/playlists/141481:Drama>, <Playlist:/playlists/362220:Reality-TV>, <Playlist:/playlists/141480:Sci-Fi>]
ERROR: Playlist IMDB: Top Rated Movies not found, so it could not be deleted. Actual playlists: [<Playlist:/playlists/282615:70's-Comedies>, <Playlist:/playlists/280507:80s-Comedy>, <Playlist:/playlists/282612:90's-Comedies>, <Playlist:/playlists/282613:2000's-Comedies>, <Playlist:/playlists/261465:All-Music>, <Playlist:/playlists/261476:All-Music>, <Playlist:/playlists/261466:Recently-Added>, <Playlist:/playlists/261475:Recently-Added>, <Playlist:/playlists/261467:Recently-Played>, <Playlist:/playlists/261477:Recently-Played>, <Playlist:/playlists/128363:Action>, <Playlist:/playlists/141517:Blake-and-Diane>, <Playlist:/playlists/372604:Blake-and-Diane-Movi>, <Playlist:/playlists/128362:Comedy>, <Playlist:/playlists/141481:Drama>, <Playlist:/playlists/362220:Reality-TV>, <Playlist:/playlists/141480:Sci-Fi>]
INFO: Updated plex watchlist in 0.0 seconds
INFO: Completed full sync in 19.3 seconds
```

**Comment by BEisem** [comment]
I tried creating new playlists on Plex called `IMDB: Top Rates Movies` and `Rotten Tomatoes: Best of 2019`but I'm getting the same errors as before...

**Comment by twolaw** [comment]
My mistake, I should have told you that syncing lists can be properly achieved only if you **sync all Plex libraries**.
Script must go through all movies and episodes to find them, do the matching and put them in plex playlists.

Reminder : you use files from https://github.com/twolaw/PlexTraktSync/tree/fix-addPlexItemToLists for this test of course.

_ERROR Playlist not found_ lines are not important here.

**Comment by BEisem** [comment]
Ah, ok... I see. I am starting a full sync now. I don't typically do full syncs because of the size of my library, but I'll let you know the results when it finishes. It will take a few hours, most likely. And yes, I am using the files at the ink you posted above for this test ...

So far, it seems to be working, or at least doing something...

```
Processing Movies  [------------------------------------]  21/11192  02:19:04INFO: Movie [3 Faces (2018)]: added to list Rotten Tomatoes: Best of 2019
Processing Movies  [------------------------------------]  65/11192  01:40:08INFO: Movie [12 Angry Men (1957)]: added to list IMDB: Top Rated Movies
Processing Movies  [------------------------------------]  76/11192  01:39:32INFO: Movie [12 Years a Slave (2013)]: added to list IMDB: Top Rated Movies
Processing Movies  [------------------------------------]  163/11192  01:30:59INFO: Movie [1917 (2019)]: added to list IMDB: Top Rated Movies
Processing Movies  [------------------------------------]  166/11192  01:30:54INFO: Movie [2001: A Space Odyssey (1968)]: added to list IMDB: Top Rated Movies
Processing Movies  [------------------------------------]  245/11192  01:27:14INFO: Movie [Ad Astra (2019)]: added to list Rotten Tomatoes: Best of 2019
```

Although I looked in Plex and I don't see those entries added to a Playlist yet. Does it add them in Plex at the end of the run?

**Comment by twolaw** [comment]
Yes, at the end.
First it collects them all, then it sends the command to plex to create the lists.

**Comment by BEisem** [comment]
I just completed a full sync, and it worked! Both playlists were added to Plex at the end of the sync.

I have a couple of suggestions. Let me know if you want me to add them as separate Feature Request issues...

The first is, I know you changed the "Playlist not found" to debug log level, and that will get rid of the errors. But it would be nice to have some sort of output at the end, indicating how many (if any) items were added to each list, or what actually happened during the playlist sync.

The other thing, if it's possible, it would be great if it would sync playlists without having to do a full sync of all libraries. If there is a way to add in this logic, let's say I sync only Movies libraries (`--sync=movies`), could it still sync movies playlists? The reason I ask is, I will probably never do a full sync of all libraries, because of the size of my libraries and the time it takes. I typically sync TV shows throughout the day, and then do a single Movie sync at night. So it would be nice if it handled the Movie playlists during that movie sync, and if there were any TV playlists, it could handle those during the TV library sync. Does that make sense?

**Comment by BEisem** [comment]
One other thing, I just noticed after running this, the trakt_cache.sqlite file has ballooned from 115.6MB to 1.96GB. Did the playlist sync really add that much more to the cache file?

**Comment by twolaw** [comment]
I don't think the cause is playlist.
I think it is due to your full sync with large libraries which heavily requests trakt API.

**Comment by BEisem** [comment]
I'm not so sure ... I've run both `--sync=tv` and `--sync=movies` using this instance, and the largest I've ever seen the cache is the previously mentioned 115.6MB. Essentially, the two combined (`--sync=tv` + `--sync=movies`) should be equivalent to a full sync, right?

But then, after running a full sync using https://github.com/Taxel/PlexTraktSync/pull/293, suddenly the cache file is almost 20x larger. It's not a big deal, because it's still a reasonable size, but I just wanted to point it out as a possible problem.

I'll try running a full sync using the main version, and see if I get the same results afterwards...

**Comment by BEisem** [comment]
I just completed a full sync using the main branch, and the `trakt-cache.sqlite` file is only 118.4MB.

**Comment by twolaw** [comment]
Ok then you can create a new issue about this cache size problem.
Give as much details as you can (version 0.9.x , step-by-step to reproduce, etc...)

I close this issue now.

**Comment by glensc** [comment]
I merged to 0.9.x cache command that can print URLs (requires requests_cache 0.6.x):
- https://github.com/Taxel/PlexTraktSync/pull/296

it may be useful giving idea of cache contents.
